mesa/agent.py

```diff
@@ -188,15 +188,21 @@ def shuffle(self, inplace: bool = False) -> AgentSet:
 
         Returns:
             AgentSet: A shuffled AgentSet. Returns the current AgentSet if inplace is True.
-        """
-        shuffled_agents = list(self)
-        self.random.shuffle(shuffled_agents)
 
-        return (
-            AgentSet(shuffled_agents, self.model)
-            if not inplace
-            else self._update(shuffled_agents)
-        )
+        Note:
+            Using inplace = True is more performant
+
+        """
+        weakrefs = list(self._agents.keyrefs())
+        self.random.shuffle(weakrefs)
+
+        if inplace:
+            self._agents.data = {entry: None for entry in weakrefs}
```

Comment:
How come in this line there is no need for `(agent := ref()) is not None)`, but it is needed in L204?

Reply:
When you create a new AgentSet, it has its own internal WeakkeyDict. Each WeakkeyDict creates its own weakrefs.

As can be seen in [the docs](https://docs.python.org/3/library/weakref.html#weakref.ref), `weakref.ref` takes an optional callback. In the case of WeakkeyDict, this callback is a method on the dict itself (i.e., `self.remove`). So, when not shuffling in place, I have to unpack the weakrefs to ensure the new WeakkeyDict behaves correctly. 

Moreover, AgentSet takes an iteratable of agents as an argument. Not an iterable of weakref. ref instances.

Hope this clarifies.